scipy/optimize/_linprog_util.py

```diff
@@ -368,84 +374,74 @@ def _clean_inputs(lp):
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": a lower bound must be less than or equal to the "
-                    "corresponding upper bound")
-            if b[0] == np.inf:
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": infinity is not a valid lower bound")
-            if b[1] == -np.inf:
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": negative infinity is not a valid upper bound")
-            lb = float(b[0]) if b[0] is not None and b[0] != -np.inf else None
-            ub = float(b[1]) if b[1] is not None and b[1] != np.inf else None
-            clean_bounds.append((lb, ub))
-        bounds = clean_bounds
-    except ValueError as e:
-        if "could not convert string to float" in e.args[0]:
-            raise TypeError
-        else:
-            raise e
-    except TypeError as e:
-        print(e)
-        raise TypeError(
-            "Invalid input for linprog: bounds must be a sequence of "
-            "(min,max) pairs, each defining bounds on an element of x ")
-
-    return _LPProblem(c, A_ub, b_ub, A_eq, b_eq, bounds, x0)
+    # Bounds can be one of these formats:
+    # (1) None
+    # (2) a sequence with 2 scalars
+    # (3) a sequence with 1 element as (2)
+    # (4) a sequence with N elements, all with 2 values (N is the size of x)
+    # (5) a sequence with 2 elements, both sequences with N values
+    # (6) a 2-D array, with shape N x 2 or 2 x N
+    # Unspecified bounds can be represented by None or (-)np.inf.
+    # All formats are converted into a N x 2 np.array with (-)np.inf where bounds are unspecified.
+    # Strings in input result in a ValueError
+    clean_bounds = np.zeros((n_x,2))
+
+    bounds_valid = False
+    # Determine shape of provided bounds
+    # np.shape returns a tuple, but only if sizes are consistent
+    bsh = np.shape(bounds)
+    
+    # 1. Check if bounds can be interpreted as n_x pairs (n_x is the number of variables)
+    #    Bounds can have sizes n_x x 2 and 2 x n_x
+    if len(bsh) == 2:
+        if bsh[0] == n_x and bsh[1] == 2: 
+            for i in range(n_x):
+                bi = bounds[i] # no need to check if length == 2, np.shape did that
+                clean_bounds[i,:] = bi
+                bounds_valid = True
+        elif bsh[0] == 2 and bsh[1] == n_x: 
+            for i in range(2):
+                bi = bounds[i] # no need to check length, np.shape did that
+                clean_bounds[:,i] = bi
+                bounds_valid = True
```

Comment:
See above. Catching dimension inconsistency first makes it more complicated but not impossible to be specific. What specificity would be needed?

Reply:
Considering your comment below (not bend over backwards), I think yesterday's update raises sufficiently specific errors in case of incorrect dimensions or incorrect input types.